include/zstd/mutex.h

```diff
@@ -0,0 +1,202 @@
+#ifndef ZEPHYR_INCLUDE_ZSTD_MUTEX_H_
+#define ZEPHYR_INCLUDE_ZSTD_MUTEX_H_
+#ifdef __cplusplus
+#ifdef CONFIG_ZSTD_MUTEX
+
+#include <kernel.h>
+
+#include <chrono>
+
+namespace zstd
+{
+
+using std::chrono::milliseconds;
+using std::chrono::duration;
+using std::chrono::duration_cast;
+using std::chrono::time_point;
+using std::chrono::steady_clock;
+
+/*******************************************************************************
+ * zstd::mutex
+ ******************************************************************************/
+class mutex
+{
+public:
+	typedef struct k_mutex *native_handle_type;
+
+	mutex();
+	virtual ~mutex() = default;
+	mutex(const mutex&) = delete;
+	mutex& operator=(const mutex&) = delete;
+
+#if defined(CONFIG_ZSTD_NEW_HEAP) || \
+   (defined(CONFIG_ZSTD_NEW_SLAB) && CONFIG_ZSTD_MUTEX_NEW_SLAB_MAX_COUNT > 0)
+	static void *operator new(size_t size);
+	static void operator delete(void *ptr);
+#endif
+
+	void lock();
+	bool try_lock();
+	void unlock();
+
+	native_handle_type native_handle();
+
+protected:
+	struct k_mutex m_mtx;
+	struct k_mutex *const m_mtx_ptr;
+}; /* class mutex */
+
+/*******************************************************************************
+ * zstd::recursive_mutex
+ ******************************************************************************/
+class recursive_mutex
+{
+public:
+	typedef struct k_mutex *native_handle_type;
+
+	recursive_mutex();
+	virtual ~recursive_mutex() = default;
+	recursive_mutex(const recursive_mutex&) = delete;
+	recursive_mutex& operator=(const recursive_mutex&) = delete;
+
+#if defined(CONFIG_ZSTD_NEW_HEAP) || \
+   (defined(CONFIG_ZSTD_NEW_SLAB) && CONFIG_ZSTD_MUTEX_NEW_SLAB_MAX_COUNT > 0)
+	void *operator new(size_t size);
+	void operator delete(void *ptr);
+#endif
+
+	void lock();
+	bool try_lock();
+	void unlock();
+
+	native_handle_type native_handle();
+
+protected:
+	struct k_mutex m_mtx;
+	struct k_mutex *const m_mtx_ptr;
+}; /* class recursive_mutex */
+
+/*******************************************************************************
+ * zstd::timed_mutex
+ ******************************************************************************/
+class timed_mutex : public mutex
+{
+public:
+	timed_mutex();
+	virtual ~timed_mutex() = default;
+	timed_mutex(const timed_mutex&) = delete;
+	timed_mutex& operator=(const timed_mutex&) = delete;
+
+#if defined(CONFIG_ZSTD_NEW_HEAP) || \
+   (defined(CONFIG_ZSTD_NEW_SLAB) && CONFIG_ZSTD_MUTEX_NEW_SLAB_MAX_COUNT > 0)
+	void *operator new(size_t size);
+	void operator delete(void *ptr);
+#endif
+
+	template <typename Rel, typename Period>
+	bool try_lock_for(const duration<Rel, Period>& rel_time)
+	{
+		if ((m_mtx_ptr->lock_count > 0) &&
+			(k_current_get() == m_mtx_ptr->owner)) {
+			return false;
```

Comment:
The Mutex C++ named requirements state that an operation that will cause deadlock should throw a `std::system_error` with code `std::errc::resource_deadlock_would_occur`.